File: list2/java_code/modfield/GFTest.java
public class GFTest {
	public static void main(String[] args) {
		GF a = new GF(p);
		GF b = new GF(p, 5);
		GF c = new GF(p, 11);
		GF d = new GF(p, 11);

		assert a.getValue() == 0;

		assert(a.compareTo(b) == -1);
		assert(b.compareTo(a) == 1);
		assert(b.compareTo(c) != 1);
		assert(c.compareTo(d) != 1);
		assert(c.compareTo(d) == 0);
		assert(a.compareTo(b) != 0);

		assert((a.add(b)).compareTo(new GF(p, 5)) == 0);
		assert((c.sub(b)).compareTo(new GF(p, 6)) == 0);
		assert((a.mul(b)).compareTo(new GF(p, 0)) == 0);
		assert((b.mul(b.inverse())).compareTo(new GF(p, 1)) == 0);
		assert(((new GF(p, 1)).div(d)).compareTo(d.inverse()) == 0);
	
		assert(a.getCharacteristic() == p);
	}

	static final long p = 1234577;
}

File: list2/rust_code/modfield/tests/gf_test.rs
use modfield::*;

const P: u64 = 1234577;

#[test]
fn test_order() {
    let a = GF::<P>::new();
    let b = GF::<P>::from_u64(5);
    let c = GF::<P>::from_u64(11);
    let d = GF::<P>::from_u64(11);

    assert!(a < b);
    assert!(b > a);
    assert!(b <= c);
    assert!(c <= d);
    assert!(c == d);
    assert!(a != b);
}

#[test]
fn test_arithmetic() {
    let a = GF::<P>::new();
    let b = GF::<P>::from_u64(5);
    let c = GF::<P>::from_u64(11);
    let d = GF::<P>::from_u64(11);

    assert_eq!(a + b, GF::<P>::from_u64(5));
    assert_eq!(c - b, GF::<P>::from_u64(6));
    assert_eq!(a * b, GF::<P>::from_u64(0));
    assert_eq!(b * b.inverse().unwrap(), GF::<P>::from_u64(1));
    assert_eq!((GF::<P>::from_u64(1) / d).unwrap(), d.inverse().unwrap());
}

#[test]
fn test_assignment() {
    let mut a = GF::<P>::new();

    a = GF::<P>::from_u64(1);
    a -= 5;
    assert_eq!(a, GF::<P>::from_u64(1234573));
    a -= 1;
    assert_eq!(a, GF::<P>::from_u64(1234572));
    a *= 2;
    assert_eq!(a, GF::<P>::from_u64(1234567));
    a /= 2;
    assert_eq!(a, GF::<P>::from_u64(1234572));
}

#[test]
fn test_other() {
    let a = GF::<P>::new();
    assert_eq!(a.get_characteristic(), P);
    assert_eq!(a.get_value(), 0);
}

